tsl/src/chunk.c

```diff
@@ -1350,43 +1456,660 @@ relation_split_info_free(RelationSplitInfo *rsi, int ti_options)
+
+	BufferHeapTupleTableSlot *hslot;
+	int routingindex = -1;
+
+	while (table_scan_getnextslot(scan, ForwardScanDirection, srcslot))
+	{
+		RelationWriteState *rws = NULL;
+		HeapTuple tuple;
+		Buffer buf;
+		bool isdead;
+		bool isalive = false;
+
+		CHECK_FOR_INTERRUPTS();
+		ResetExprContext(econtext);
+
+		tuple = ExecFetchSlotHeapTuple(srcslot, false, NULL);
+
+		if (TTS_IS_ARROWTUPLE(srcslot))
+		{
+			ArrowTupleTableSlot *aslot = (ArrowTupleTableSlot *) srcslot;
+			hslot = (BufferHeapTupleTableSlot *) aslot->child_slot;
+		}
+		else
+		{
+			hslot = (BufferHeapTupleTableSlot *) srcslot;
+		}
+
+		buf = hslot->buffer;
+
+		LockBuffer(buf, BUFFER_LOCK_SHARE);
+
+		switch (HeapTupleSatisfiesVacuum(tuple, scontext->cutoffs.OldestXmin, buf))
+		{
+			case HEAPTUPLE_DEAD:
+				/* Definitely dead */
+				isdead = true;
+				break;
+			case HEAPTUPLE_RECENTLY_DEAD:
+				tups_recently_dead += 1;
+				isdead = false;
+				break;
+			case HEAPTUPLE_LIVE:
+				/* Live or recently dead, must copy it */
+				isdead = false;
+				isalive = true;
+				break;
+			case HEAPTUPLE_INSERT_IN_PROGRESS:
+				/*
+				 * Since we hold exclusive lock on the relation, normally the
+				 * only way to see this is if it was inserted earlier in our
+				 * own transaction. Give a warning if this case does not
+				 * apply; in any case we better copy it.
+				 */
+				if (!TransactionIdIsCurrentTransactionId(HeapTupleHeaderGetXmin(tuple->t_data)))
+					elog(WARNING,
+						 "concurrent insert in progress within table \"%s\"",
+						 RelationGetRelationName(srcrel));
+				/* treat as live */
+				isdead = false;
+				isalive = true;
+				break;
+			case HEAPTUPLE_DELETE_IN_PROGRESS:
+				/*
+				 * Similar situation to INSERT_IN_PROGRESS case.
+				 */
+				if (!TransactionIdIsCurrentTransactionId(
+						HeapTupleHeaderGetUpdateXid(tuple->t_data)))
+					elog(WARNING,
+						 "concurrent delete in progress within table \"%s\"",
+						 RelationGetRelationName(srcrel));
+				/* treat as recently dead */
+				tups_recently_dead += 1;
+				isalive = true;
+				isdead = false;
+				break;
+			default:
+				elog(ERROR, "unexpected HeapTupleSatisfiesVacuum result");
+				isdead = false; /* keep compiler quiet */
+				break;
+		}
+
+		LockBuffer(buf, BUFFER_LOCK_UNLOCK);
+
+		/*
+		 * Route the tuple to the matching (new) partition. The routing is
+		 * done in a loop because compressed tuple segments might be split
+		 * into multiple sub-segment tuples if the split is in the middle of
+		 * that segment.
+		 */
+		while ((tuple = sp->route_next_tuple(srcslot, scontext, &routingindex)))
+		{
+			Assert(routingindex >= 0 && routingindex < scontext->split_factor);
+			rws = &scontext->rws[routingindex];
+
+			if (isdead)
+			{
+				tups_vacuumed += 1;
+				/* heap rewrite module still needs to see it... */
+				if (rewrite_heap_dead_tuple(rws->rwstate, tuple))
+				{
+					/* A previous recently-dead tuple is now known dead */
+					tups_vacuumed += 1;
```

Comment:
Not exactly sure how to get this covered by a test. This code is modeled on the cluster code in PG, so should be correct.

I might try an isolation test. Probably one where there are concurrent transactions that the finish before it gets here.